src/pandablocks_ioc/ioc.py

```diff
@@ -175,6 +177,41 @@ def create_softioc(
             asyncio.run_coroutine_threadsafe(client.close(), dispatcher.loop).result()
 
 
+async def get_panda_ver_info(
+    client: AsyncioClient,
+) -> tuple[str]:
+    """Function that gets version information from the PandA using the IDN command
+
+    Args:
+        client (AsyncioClient): Client used for commuication with the PandA
+
+    Returns:
+        Dict[str, str]: Dictionary mapping firmware name to version
+    """
+
+    idn = await client.send(Get("*IDN"))
+
+    # Currently, IDN reports sw, fpga, and rootfs versions
+    firmware_versions = {"PandA SW": "Unknown", "FPGA": "Unknown", "rootfs": "Unknown"}
+
+    prev_key = None
+    firmware_list = list(firmware_versions.keys())
+    for i, key in enumerate(list(firmware_list)):
+        try:
+            # Start of string to next key name is version of previous key
+            if prev_key is not None:
+                firmware_versions[prev_key] = idn.split(f"{key}: ")[0]
+            idn = idn.split(f"{key}: ")[1]
+            prev_key = key
+            # If we've gone through the list, last key's version is remaining string
+            if (i + 1) == len(firmware_list):
+                firmware_versions[key] = idn
+        except IndexError:
+            logging.warning(f"Failed to get {key} version information!")
+
+    return firmware_versions
```

Comment:
```suggestion
    # Currently, IDN reports sw, fpga, and rootfs versions
    firmware_versions = {"PANDA_SW": "Unknown", "FPGA": "Unknown", "ROOTFS": "Unknown"}
    pattern = re.compile(r"([^:]+):\s*([^:]+?)(?=\s+\w+:\s|$)")
    matches = {
        name.strip().replace(" ", "_").upper(): value.strip()
        for name, value in pattern.findall(idn)
    }
    for name in firmware_versions:
        if name not in matches:
            logging.warning(f"Failed to get {name} version information!")
    return firmware_versions.update(matches)
```

For reference in your testing the regex does this:

``` python
import re


def unpack_string(s):
    # Use regex to find all NAME: VALUE pairs
    pattern = re.compile(r"([^:]+):\s*([^:]+?)(?=\s+\w+:\s|$)")
    matches = pattern.findall(s)

    # Create the dictionary
    result = {name.strip(): value.strip() for name, value in matches}

    return result


s = "NAME1: VALUE11 VALUE12 NAME2: VALUE21 NAME3: VALUE31 VALUE32 VALUE33"
result = unpack_string(s)
# {'NAME1': 'VALUE11 VALUE12', 'NAME2': 'VALUE21', 'NAME3': 'VALUE31 VALUE32 VALUE33'}
print(result)

s = "PandA SW: 330bd94-dirty FPGA: 0.1.9 d1275f61 00000000"
result = unpack_string(s)
# {'PandA SW': '330bd94-dirty', 'FPGA': '0.1.9 d1275f61 00000000'}
print(result)
```

Reply:
Thinking about this now, we should probably instead make the names explicit. It's annoying the message from the server can have indices with spaces between, but spaces are also the seperator.
```suggestion
    # Currently, IDN reports sw, fpga, and rootfs versions
    firmware_versions = {"PandA SW": "Unknown", "FPGA": "Unknown", "rootfs": "Unknown"}
    
    # raise error if len(idn).split(":")/2 != len(firmware_versions)

    for name in firmware_versions:
        pattern = re.compile(
            rf'{re.escape(key)}:\s*([^:]+?)(?=\s*\b(?:{"|".join(map(re.escape, firmware_versions))}):|$)'
        )
        if (match := pattern.search(s)):
            firmware_versions[name] = match.group(1).strip()
        else
            logging.warning(f"Failed to get {name} version information!")

    return {
        EpicsName(name.upper().replace(" ", "_")):
        version
        for name, version in firmware_versions.items()
    }
```

Test:
``` python
import re


def unpack_string(s, keys: dict[str, str], separator: str = ":") -> dict:
    for key in keys:
        pattern = re.compile(
            rf'{re.escape(key)}{separator}\s*([^:]+?)(?=\s*\b(?:{"|".join(map(re.escape, keys))}){separator}|$)'
        )
        match = pattern.search(s)
        if match:
            keys[key] = match.group(1).strip()
        else:
            print("warning")
    return keys


# Test cases
s1 = "NAME1: VALUE11 VALUE12 NAME2: VALUE21 NAME3: VALUE31 VALUE32 VALUE33"
print(unpack_string(s1, {"NAME1": "A", "NAME2": "A", "NAME3": "A", "NAME4": "A"}))
# Gives {'NAME1': 'VALUE11 VALUE12', 'NAME2': 'VALUE21', 'NAME3': 'VALUE31 VALUE32 VALUE33'} and "warning"

s2 = "PandA SW: 330bd94-dirty FPGA: 0.1.9 d1275f61 00000000"
print(unpack_string(s2, {"PandA SW": "A", "FPGA": "1 2 3"}))
# Gives {'PandA SW': '330bd94-dirty', 'FPGA': '0.1.9 d1275f61 00000000'}
```